File: backend/local_client.js
/**
 * 로컬 PC에서 실행 — Tapo RTSP → EC2 전송
 * 실행: node local_client.js
 */

const ffmpeg = require('fluent-ffmpeg');
const ffmpegPath = require('ffmpeg-static');
const fetch = require('node-fetch');
const FormData = require('form-data');
const fs = require('fs');
const path = require('path');

ffmpeg.setFfmpegPath(ffmpegPath);

const EC2_URL = 'http://3.34.177.253:8000';
const RTSP_URL = 'rtsp://tapo1234:123456788@10.17.239.85/stream1';
const INTERVAL_MS = 5000;  // 5초마다 캡처

// 좌석 ROI (x, y, width, height) — reference.jpg 확인 후 수정
// 일단은 전체 화면을 좌석 1개로 전송 (테스트용)
const SEAT_ROIS = [
  { seat_id: 1, x: 0,   y: 0,   w: 640, h: 360 },
  { seat_id: 2, x: 640, y: 0,   w: 640, h: 360 },
  { seat_id: 3, x: 0,   y: 360, w: 640, h: 360 },
  { seat_id: 4, x: 640, y: 360, w: 640, h: 360 },
];

const TMP_DIR = path.join(__dirname, 'tmp_frames');
if (!fs.existsSync(TMP_DIR)) fs.mkdirSync(TMP_DIR);


async function captureFrame(outputPath) {
  return new Promise((resolve, reject) => {
    ffmpeg(RTSP_URL)
      .inputOptions(['-rtsp_transport', 'tcp'])
      .frames(1)
      .output(outputPath)
      .on('end', resolve)
      .on('error', reject)
      .run();
  });
}


async function cropAndSend(fullFramePath, roi) {
  const cropPath = path.join(TMP_DIR, `seat_${roi.seat_id}.jpg`);

  // ROI 크롭
  await new Promise((resolve, reject) => {
    ffmpeg(fullFramePath)
      .videoFilter(`crop=${roi.w}:${roi.h}:${roi.x}:${roi.y}`)
      .frames(1)
      .output(cropPath)
      .on('end', resolve)
      .on('error', reject)
      .run();
  });

  // EC2로 전송
  const form = new FormData();
  form.append('file', fs.createReadStream(cropPath), 'frame.jpg');

  try {
    const res = await fetch(`${EC2_URL}/analyze?seat_id=${roi.seat_id}`, {
      method: 'POST',
      body: form,
      headers: form.getHeaders(),
      timeout: 15000,
    });
    const data = await res.json();
    console.log(`  좌석 ${roi.seat_id}: ${data.status} | 사람=${data.has_person} 물건=${data.has_items}`);
  } catch (e) {
    console.log(`  좌석 ${roi.seat_id} 전송 실패: ${e.message}`);
  }
}


async function checkEC2() {
  try {
    const res = await fetch(`${EC2_URL}/health`, { timeout: 5000 });
    const data = await res.json();
    console.log('EC2 연결 확인:', data);
    return true;
  } catch (e) {
    console.error('EC2 연결 실패:', e.message);
    return false;
  }
}


async function loop() {
  const fullFramePath = path.join(TMP_DIR, 'full_frame.jpg');

  console.log(`[${new Date().toLocaleTimeString()}] 프레임 캡처 중...`);

  try {
    await captureFrame(fullFramePath);

    // 첫 실행 시 reference.jpg 저장
    const refPath = path.join(__dirname, 'reference.jpg');
    if (!fs.existsSync(refPath)) {
      fs.copyFileSync(fullFramePath, refPath);
      console.log(`참조 이미지 저장됨 → ${refPath}`);
      console.log('이 이미지를 열어 좌석 ROI 좌표를 확인하고 local_client.js의 SEAT_ROIS를 수정하세요.');
    }

    for (const roi of SEAT_ROIS) {
      await cropAndSend(fullFramePath, roi);
    }
  } catch (e) {
    console.error('캡처 실패:', e.message);
  }

  setTimeout(loop, INTERVAL_MS);
}


(async () => {
  console.log('=== Tapo → EC2 클라이언트 시작 ===');
  console.log(`RTSP: ${RTSP_URL}`);
  console.log(`EC2:  ${EC2_URL}\n`);

  const ok = await checkEC2();
  if (!ok) process.exit(1);

  await loop();
})();

File: backend/local_client.py
"""
로컬 PC에서 실행 (Tapo 카메라와 같은 네트워크)
설치: pip install opencv-python requests
실행: python local_client.py
"""
import cv2
import requests
import time
import sys
import os

EC2_URL = "http://13.239.60.158:8000"
INTERVAL = 5  # 몇 초마다 분석할지

RTSP_CANDIDATES = [
    "rtsp://tapo1234:123456788@10.17.239.85/stream1",
    "rtsp://tapo1234:123456788@10.91.195.85/stream1",
]


def find_rtsp():
    for url in RTSP_CANDIDATES:
        print(f"연결 시도: {url}")
        cap = cv2.VideoCapture(url)
        if cap.isOpened():
            ret, _ = cap.read()
            if ret:
                print(f"연결 성공: {url}")
                return cap
        cap.release()
    return None


def save_reference(frame):
    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "reference.jpg")
    cv2.imwrite(path, frame)
    print(f"참조 이미지 저장됨 → {path}")


def detect_person(frame):
    _, img_encoded = cv2.imencode(".jpg", frame)
    try:
        resp = requests.post(
            f"{EC2_URL}/detect",
            files={"file": ("frame.jpg", img_encoded.tobytes(), "image/jpeg")},
            timeout=20,
        )
        if resp.ok:
            d = resp.json()
            detected_labels = [f"{x['label']}({x['confidence']})" for x in d.get("detected", [])]
            label_str = ", ".join(detected_labels) if detected_labels else "없음"
            print(f"  사람 감지: {d['has_person']} | 감지된 객체: {label_str}")
        else:
            print(f"  오류: {resp.status_code} {resp.text}")
    except Exception as e:
        print(f"  전송 실패: {e}")


def main():
    try:
        r = requests.get(f"{EC2_URL}/health", timeout=5)
        print(f"EC2 연결 확인: {r.json()}")
    except Exception as e:
        print(f"EC2 연결 실패: {e}")
        sys.exit(1)

    cap = find_rtsp()
    if cap is None:
        print("RTSP 연결 실패. 카메라 IP와 자격증명을 확인하세요.")
        sys.exit(1)

    frame_count = 0
    print(f"\n사람 감지 시작 (매 {INTERVAL}초마다)\n")

    while True:
        ret, frame = cap.read()
        if not ret:
            print("프레임 읽기 실패. 재연결...")
            cap.release()
            time.sleep(3)
            cap = find_rtsp()
            if cap is None:
                break
            continue

        if frame_count == 0:
            save_reference(frame)

        print(f"[{time.strftime('%H:%M:%S')}] 분석 중...")
        detect_person(frame)

        frame_count += 1
        time.sleep(INTERVAL)

    cap.release()


if __name__ == "__main__":
    main()
